### PR TITLE
[WIP] Add PHP Bytecode support

### PR DESCRIPTION
**Pull request by mikenz**
Requires:
- PHP 7.x
- the PHP VLD extension to be installed and enabled: https://github.com/derickr/vld 

Todo:
- [x] Test with php 7.0
- [x] Test with php 7.1
- [x] Test with php 7.2
- [x] Test with php 7.3 (php-ast and php-vld plugins not yet supported on php-master)
- [x] Add AST dumping
- [x] Clean up bytecode format/layout - make fake "Intel" view?
- [ ] Test if VLD php extension is available before attempting to use it
- [ ] Allow execution?
- [ ] Add tests
- [x] Fix zend optimizer not running when a child process of nodejs

### CODE REVIEW

**Comment by mikenz** [comment]
How to build PHP versions: on Ubuntu 16.04. 

```
sudo apt-get install -y \
    build-essential \
    autoconf \
    bison \
    cmake \
    libsodium-dev \
    libxml2-dev \
    libbz2-dev \
    libmcrypt-dev \
    libcurl4-openssl-dev \
    libltdl-dev \
    libpng12-dev \
    libpspell-dev \
    libreadline-dev \
    libicu-dev \
    libxml2-dev \
    libpng-dev \
    libmcrypt-dev \
    libfreetype6 \
    libfreetype6-dev \
    imagemagick \
    libmagickwand-dev \
    zlib1g-dev \
    libssl-dev

# Checkout PHP
git clone https://github.com/php/php-src.git
cd php-src

# Build PHP 7.2.0
mkdir /opt/compiler-explorer/php-7.2.0
git checkout PHP-7.2.0
./buildconf --force
./configure --prefix=/opt/compiler-explorer/php-7.2.0 \
                  --with-config-file-path=/opt/compiler-explorer/php-7.2.0/etc \
                  --with-config-file-scan-dir=/opt/compiler-explorer/php-7.2.0/etc/conf.d \
                  --disable-cgi \
                  --disable-fpm \
                  --disable-phpdbg \
                  --disable-debug \
                  --enable-bcmath \
                  --enable-filter \
                  --enable-intl \
                  --enable-mysqlnd \
                  --enable-opcache \
                  --enable-simplexml \
                  --enable-xmlreader \
                  --enable-xmlwriter \
                  --enable-zip \
                  --with-bz2 \
                  --with-curl \
                  --with-freetype-dir \
                  --with-gd \
                  --with-jpeg-dir \
                  --with-mysql-sock=/var/run/mysqld/mysqld.sock \
                  --with-mysqli \
                  --with-openssl \
                  --with-pdo-mysql=mysqlnd \
                  --with-pdo-sqlite \
                  --with-png-dir \
                  --with-sqlite3 \
                  --with-sodium \
                  --enable-soap \
                  --with-zlib
make -j4
make install
mkdir /opt/compiler-explorer/php-7.2.0/etc/conf.d
echo "zend_extension=opcache.so" > /opt/compiler-explorer/php-7.2.0/etc/conf.d/10-opcache.ini

# VLD extension for php 7.2.0
cd
git clone https://github.com/derickr/vld.git
cd vld
/opt/compiler-explorer/php-7.2.0/bin/phpize
./configure --with-php-config=/opt/compiler-explorer/php-7.2.0/bin/php-config
echo "extension=vld" > /opt/compiler-explorer/php-7.2.0/etc/conf.d/20-vld.ini

# Build PHP 7.1.12
mkdir /opt/compiler-explorer/php-7.1.12
cd ~/php-src
git checkout PHP-7.1.12
make clean
./buildconf --force
./configure --prefix=/opt/compiler-explorer/php-7.1.12 \
                  --with-config-file-path=/opt/compiler-explorer/php-7.1.12/etc \
                  --with-config-file-scan-dir=/opt/compiler-explorer/php-7.1.12/etc/conf.d \
                  --disable-cgi \
                  --disable-fpm \
                  --disable-phpdbg \
                  --disable-debug \
                  --enable-bcmath \
                  --enable-filter \
                  --enable-intl \
                  --enable-mysqlnd \
                  --enable-opcache \
                  --enable-simplexml \
                  --enable-xmlreader \
                  --enable-xmlwriter \
                  --enable-zip \
                  --with-bz2 \
                  --with-curl \
                  --with-freetype-dir \
                  --with-gd \
                  --with-jpeg-dir \
                  --with-mysql-sock=/var/run/mysqld/mysqld.sock \
                  --with-mysqli \
                  --with-openssl \
                  --with-pdo-mysql=mysqlnd \
                  --with-pdo-sqlite \
                  --with-png-dir \
                  --with-sqlite3 \
                  --with-mcrypt \
                  --enable-soap \
                  --with-zlib
make -j4
make install
mkdir /opt/compiler-explorer/php-7.1.12/etc/conf.d
echo "zend_extension=opcache.so" > /opt/compiler-explorer/php-7.1.12/etc/conf.d/10-opcache.ini

# VLD extension for php 7.1.12
cd ~/vld
make clean
rm configure.ac
/opt/compiler-explorer/php-7.1.12/bin/phpize
./configure --with-php-config=/opt/compiler-explorer/php-7.1.12/bin/php-config
make
make install
echo "extension=vld.so" > /opt/compiler-explorer/php-7.1.12/etc/conf.d/20-vld.ini

# Build PHP 7.0.26
mkdir /opt/compiler-explorer/php-7.0.26
cd ~/php-src
git checkout PHP-7.0.26
make clean
./buildconf --force
./configure --prefix=/opt/compiler-explorer/php-7.0.26 \
                  --with-config-file-path=/opt/compiler-explorer/php-7.0.26/etc \
                  --with-config-file-scan-dir=/opt/compiler-explorer/php-7.0.26/etc/conf.d \
                  --disable-cgi \
                  --disable-fpm \
                  --disable-phpdbg \
                  --disable-debug \
                  --enable-bcmath \
                  --enable-filter \
                  --enable-intl \
                  --enable-mysqlnd \
                  --enable-opcache \
                  --enable-simplexml \
                  --enable-xmlreader \
                  --enable-xmlwriter \
                  --enable-zip \
                  --with-bz2 \
                  --with-curl \
                  --with-freetype-dir \
                  --with-gd \
                  --with-jpeg-dir \
                  --with-mysql-sock=/var/run/mysqld/mysqld.sock \
                  --with-mysqli \
                  --with-openssl \
                  --with-pdo-mysql=mysqlnd \
                  --with-pdo-sqlite \
                  --with-png-dir \
                  --with-sqlite3 \
                  --with-mcrypt \
                  --enable-soap \
                  --with-zlib
make -j4
make install
mkdir /opt/compiler-explorer/php-7.0.26/etc/conf.d
echo "zend_extension=opcache.so" > /opt/compiler-explorer/php-7.0.26/etc/conf.d/10-opcache.ini

# VLD extension for php 7.0.26
cd ~/vld
make clean
rm configure.ac
/opt/compiler-explorer/php-7.0.26/bin/phpize
./configure --with-php-config=/opt/compiler-explorer/php-7.0.26/bin/php-config
make
make install
echo "extension=vld.so" > /opt/compiler-explorer/php-7.0.26/etc/conf.d/20-vld.ini

```

File: lib/compilers/php.js [hunk 1]
```diff
@@ -74,6 +74,14 @@ function compilePHP(info, env) {
         });
     };
 
+    compiler.generateAST = function (inputFilename, options) {
+        let newOptions = ['etc/scripts/dump-php-ast.php', inputFilename];
+        let execOptions = this.getDefaultExecOptions();
+        return this.exec(this.compiler.exe, newOptions, execOptions).then(function (result) {
```
Comment (RabsRincon) [inline lib/compilers/php.js]: `.then(result => result.stdout)` is also a good one liner for this

File: etc/config/builtin.php.properties [hunk 2]
```diff
@@ -0,0 +1,2 @@
+sourcepath=./examples/php/
```
Comment (RabsRincon) [inline etc/config/builtin.php.properties]: File not needed anymore. Info now stored in `libs/languages.js`

File: etc/config/php.amazon.properties [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Default settings for PHP
```
Comment (RabsRincon) [inline etc/config/php.amazon.properties]: (Amazon)

File: examples/php/Functions_in_functions.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    function b() {
+        function c() {
+            return "c";
+        }
+        return "b" . c();
+    }
+    return "a" . b();
+};
+
+echo a(); // returns 'abc'
```
Comment (RabsRincon) [inline examples/php/Functions_in_functions.php]: Newline at eof missing

File: etc/config/php.amazon.properties [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+compilers=&php7
+defaultCompiler=php720
+
+group.php7.compilers=phpmaster:php720:php7112:php7026
+compiler.phpmaster.exe=/opt/compiler-explorer/php-master/bin/php
+compiler.phpmaster.alias=/usr/bin/php-master
+compiler.phpmaster.name=php master
```
Comment (RabsRincon) [inline etc/config/php.amazon.properties]: php name should probably be uppercase to match the rest of the languages

Also, this specifical version, to match the rest of the compilers versionings, should get the `master` changed to `(trunk)`

File: lib/compilers/php.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        // Only allow "-e" and enabling the opcode cache option, everthing else looks dangerous
+        userOptions = (userOptions || []).filter(option => option == '-e' || option == '-dopcache.enable_cli=1');
+
+        // Hack: Pass output filename as the last option
+        return options.concat(userOptions || []).concat([this.filename(inputFilename), outputFilename]);
+    };
+
+    compiler.runCompiler = function (compiler, options, inputFilename, execOptions) {
+        // Hack: Get output filename from last option
```
Comment (RabsRincon) [inline lib/compilers/php.js]: Is there a better way to do this? Seems a bit sketchy

File: test/cases/php-functions-in-functions.asm [hunk 7]
```diff
@@ -0,0 +1,63 @@
+	Findingentrypoints
```
Comment (RabsRincon) [inline test/cases/php-functions-in-functions.asm]: Thanks for the tests!

File: lib/base-compiler.js [hunk 8]
```diff
@@ -39,13 +39,14 @@ const child_process = require('child_process'),
 
 function Compile(compiler, env) {
```
Comment (RabsRincon) [inline lib/base-compiler.js]: Bad merge

File: lib/compilers/pascal.js [hunk 9]
```diff
@@ -65,11 +65,11 @@ function compileFPC(info, env) {
     };
 
     compiler.getOutputFilename = function (dirPath, outputFilebase) {
```
Comment (RabsRincon) [inline lib/compilers/pascal.js]: Bad merge

**Comment by RabsRincon** [comment]
Hi! As part of our plans for February, we are trying to reduce our number of long standing pull requests. We would like to know if you are planning on working on this, or if we should try to finish it for you (Or close it in case we are unable to).

We don't want to push you into doing something you might not want to do/don't have the time right now, so we'll understand if you choose not to work on this. No hard feelings :). We are indeed grateful you took the time to try to improve CE.

**Comment by RabsRincon** [comment]
Hi! I'm going to close this PR for the lack of activity. (No hard feelings, we're just cleaning our PR list up)

Please, reopen this if (Hopefuly "when"!) you get some time to finish this off, or if someone else wants to pick up the feature

:)

**Pull request closed without merging**